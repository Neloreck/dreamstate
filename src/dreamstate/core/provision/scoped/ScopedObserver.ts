import {
  createElement,
  DispatchWithoutAction,
  ReactChild,
  ReactElement,
  useEffect,
  useMemo,
  useReducer
} from "react";

import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { forceUpdateReducer } from "@/dreamstate/core/utils/forceUpdateReducer";
import { TAnyContextManagerConstructor, TAnyObject, TCallable } from "@/dreamstate/types";

export interface IScopedObserverProps<T> {
  children?: ReactChild;
  ManagerClass: TAnyContextManagerConstructor;
  scope: IScopeContext;
  initialState?: T;
  dependencies: Array<TAnyContextManagerConstructor>;
}

/**
 * ScopedObserver component that does data provision for specific source in an isolated react node.
 * Receives current scope information and specific manager reference for provision.
 */
export function ScopedObserver<T>({
  children,
  initialState,
  dependencies,
  ManagerClass,
  scope
}: IScopedObserverProps<T>): ReactElement {
  const reducer: [ TAnyObject | null, DispatchWithoutAction ] = useReducer(forceUpdateReducer, null);

  /**
   * Use memo for first and single init of required components.
   * The point is to force registering before provider rendering for first init.
   *
   * Registering will work in the same way even if it is called multiple times.
   * Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useMemo(function(): void {
    scope.INTERNAL.registerService(ManagerClass, initialState);
  }, dependencies);

  /**
   * Mount current observers and trigger related lifecycle methods when needed.
   * Count references of providers to detect whether we start provisioning or ending it.
   *
   * ! Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useEffect(function(): TCallable {
    const onUpdateNeeded: TCallable = reducer[1];

    scope.INTERNAL.registerService(ManagerClass, initialState);
    scope.INTERNAL.addServiceObserver(ManagerClass, onUpdateNeeded);

    return function(): void {
      scope.INTERNAL.removeServiceObserver(ManagerClass, onUpdateNeeded);
    };
  }, dependencies);

  /**
   * Create simple react provider based on current ManagerClass and provide data from scope registry.
   */
  return createElement(
    ManagerClass.REACT_CONTEXT.Provider,
    { value: scope.INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY.get(ManagerClass) },
    children
  );
}
