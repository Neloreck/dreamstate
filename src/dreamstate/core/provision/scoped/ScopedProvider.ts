import { createElement, ReactElement, ReactNode, useEffect, useMemo } from "react";

import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { useForceUpdate } from "@/dreamstate/core/utils/useForceUpdate";
import { TAnyContextManagerConstructor, TAnyObject, TCallable } from "@/dreamstate/types";

export interface IScopedObserverProps<T> {
  ManagerClass: TAnyContextManagerConstructor;
  children?: ReactNode;
  dependencies: Array<TAnyContextManagerConstructor>;
  initialState?: T;
  scope: IScopeContext;
}

/**
 * A component that provides data for a specific context manager within an isolated React node.
 *
 * This component is responsible for observing changes within a specific scope and providing the
 * context data for the given context manager (`ManagerClass`). It allows for isolated data provisioning
 * and is useful for managing specific parts of the application state while preventing unnecessary
 * re-renders of unrelated components.
 *
 * @template T - The type of the context data to be provided by the `ManagerClass`.
 * @param {object} props - The props for the component.
 * @param {ReactNode} props.children - The child components that will receive the provided context data.
 * @param {TAnyObject | undefined} props.initialState - The initial state to be provided for the context manager.
 * @param {Array<unknown>} props.dependencies - The dependencies used to determine when the component should
 *   re-render, typically when specific context data changes.
 * @param {TAnyContextManagerConstructor} props.ManagerClass - The context manager class reference that will provide
 *   the data to the component.
 * @param {string} props.scope - The scope information that defines the context and its relationship to other parts
 *   of the application.
 * @returns {ReactElement} A React element that provides the context data for the specified `ManagerClass`
 *   within the defined scope.
 */
export function ScopedProvider<T extends TAnyObject>({
  children,
  initialState,
  dependencies,
  ManagerClass,
  scope,
}: IScopedObserverProps<T>): ReactElement {
  const forceUpdate: TCallable = useForceUpdate();

  /*
   * Use memo for first and single init of required components.
   * The point is to force registering before provider rendering for first init.
   *
   * Registering will work in the same way even if it is called multiple times.
   * Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useMemo(function(): void {
    scope.INTERNAL.registerService(ManagerClass, initialState);
  }, dependencies);

  /*
   * Mount current observers and trigger related lifecycle methods when needed.
   * Count references of providers to detect whether we start provisioning or ending it.
   *
   * ! Dependencies array is mostly used for HMR updates to force reloading on class reference changes.
   */
  useEffect(function(): TCallable {
    const isRegistered: boolean = scope.INTERNAL.registerService(ManagerClass, initialState);

    scope.INTERNAL.addServiceObserver(ManagerClass, forceUpdate);

    /*
     * Re-sync scope provider if it was registered.
     * Normally happens with HMR chunks exchange and caused problems that use react context without subscription.
     * Required to force render of subscribed components after HMR with newest fast-refresh plugins.
     */
    if (isRegistered) {
      forceUpdate();
    }

    return function(): void {
      scope.INTERNAL.removeServiceObserver(ManagerClass, forceUpdate);
    };
  }, dependencies);

  /*
   * Create simple react provider based on current ManagerClass and provide data from scope registry.
   */
  return createElement(
    ManagerClass.REACT_CONTEXT.Provider,
    {
      value: scope.INTERNAL.REGISTRY.CONTEXT_STATES_REGISTRY.get(ManagerClass),
    },
    children
  );
}

/*
 * Easier devtools usage for dev environment.
 */
if (IS_DEV) {
  ScopedProvider.displayName = "Dreamstate.ScopedProvider";
}
