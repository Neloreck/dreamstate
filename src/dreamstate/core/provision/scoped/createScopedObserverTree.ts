import { createElement, ReactNode } from "react";

import { ScopedObserver } from "@/dreamstate/core/provision/scoped/ScopedObserver";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Recursively creates React elements within the current node scope for the provided context managers.
 *
 * This function generates React elements by recursively creating observers for each context manager
 * in the `sources` array, wrapping them in their respective context providers. It ensures that the
 * React tree is rendered within the current observation and data provision scope, allowing for
 * scoped data management across the component tree.
 *
 * @template T - The type of the context data being observed and provided.
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references
 *   that will be used to create the observer tree.
 * @param {IProviderProps<T>} props - The properties of the current parent observer, which include
 *   context data, dependencies, and other necessary configurations.
 * @param {IScopeContext} scope - The current scope context used for observation and data provisioning.
 * @param {number} [current=0] - The current iteration index for the recursive rendering process.
 * @returns {ReactNode} A React node representing the recursively created observer tree for the given
 *   context managers within the specified scope.
 */
export function createScopedObserverTree<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  scope: IScopeContext,
  current: number = 0
): ReactNode {
  /*
   * Wrap children node with observers until all source observer elements are created.
   */
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      {
        ManagerClass: sources[current],
        initialState: props.initialState,
        dependencies: sources,
        scope,
      },
      createScopedObserverTree(sources, props, scope, current + 1)
    );
}
