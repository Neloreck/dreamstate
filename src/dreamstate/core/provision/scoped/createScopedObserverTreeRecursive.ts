import { createElement, ReactNode } from "react";

import { ScopedObserver } from "@/dreamstate/core/provision/combined/ScopedObserver";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Subtree provider as global scope helper.
 * Recursive impl.
 */
export function createScopedObserverTreeRecursive<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>,
  props: IProviderProps<T>,
  scope: IScopeContext,
  current: number = 0
): ReactNode {
  return current >= sources.length
    ? props.children
    : createElement(
      ScopedObserver,
      { manager: sources[current], initialState: props.initialState, dependencies: sources, scope },
      createScopedObserverTreeRecursive(sources, props, scope,current + 1)
    );
}
