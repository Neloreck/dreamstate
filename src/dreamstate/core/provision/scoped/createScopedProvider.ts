import { FunctionComponent, ReactNode, useContext } from "react";

import { log } from "@/macroses/log.macro";

import { createScopedProviderTree } from "@/dreamstate/core/provision/scoped/createScopedProviderTree";
import { IScopeContext, ScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TAnyContextManagerConstructor, TAnyValue } from "@/dreamstate/types";
import { IProviderProps } from "@/dreamstate/types/provision";

/**
 * Creates a provider that bundles all observers into scoped nodes, ensuring minimal re-renders.
 *
 * This function generates a React provider component that groups all specified context managers into scoped
 * nodes. It efficiently updates only the affected provider when changes occur, preventing unnecessary re-renders
 * for the entire provider tree. This optimization helps improve performance by limiting updates to only the
 * parts of the tree that require them.
 *
 * @template T - The type of the provider props.
 * @param {Array<TAnyContextManagerConstructor>} sources - An array of context manager class references
 *   that should be provided as context within scoped nodes.
 * @returns {FunctionComponent<IProviderProps<T>>} A React function component that acts as a scoped provider
 *   for the specified context manager classes, optimizing re-renders.
 */
export function createScopedProvider<T extends IProviderProps<TAnyValue>>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Provider(props: T): ReactNode {
    const scope: IScopeContext = useContext(ScopeContext);

    /*
     * Warn if current observer is mounted out of scope in dev mode.
     */
    if (IS_DEV) {
      if (!scope) {
        log.error("Dreamstate providers should be used in a scope. Wrap your component tree with ScopeProvider");
      }
    }

    return createScopedProviderTree(sources, props, scope);
  }

  /*
   * Single wrapper, many observers + providers in isolated scopes.
   */
  if (IS_DEV) {
    Provider.displayName = `Dreamstate.ScopedProviders[${sources.map(function(
      it: TAnyContextManagerConstructor
    ): string {
      return it.name;
    })}]`;
  }

  return Provider as FunctionComponent<IProviderProps<T>>;
}
