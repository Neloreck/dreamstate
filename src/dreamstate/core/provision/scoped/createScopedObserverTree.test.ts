import { shallow } from "enzyme";
import { createElement } from "react";

import { createScopedObserverTree } from "@/dreamstate/core/provision/scoped/createScopedObserverTree";
import { NestedContextManager, TestContextManager } from "@/fixtures";

describe("createScopedObserverTree rendering", () => {
  it("Should match snapshot", () => {
    const recursive = shallow(
      createElement(
        "div",
        {},
        createScopedObserverTree(
          [
            TestContextManager,
            NestedContextManager,
            TestContextManager
          ],
          { children: createElement("div", {}, "bottom") },
          null as any
        )
      )
    );

    expect(recursive).toMatchSnapshot();
  });
});
