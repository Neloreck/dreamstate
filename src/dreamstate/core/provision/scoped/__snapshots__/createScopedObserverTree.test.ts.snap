// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createScopedObserverTree rendering Should match snapshot 1`] = `
<div>
  <ScopedObserver
    ManagerClass={[Function]}
    dependencies={
      Array [
        [Function],
        [Function],
        [Function],
      ]
    }
    scope={null}
  >
    <ScopedObserver
      ManagerClass={[Function]}
      dependencies={
        Array [
          [Function],
          [Function],
          [Function],
        ]
      }
      scope={null}
    >
      <ScopedObserver
        ManagerClass={[Function]}
        dependencies={
          Array [
            [Function],
            [Function],
            [Function],
          ]
        }
        scope={null}
      >
        <div>
          bottom
        </div>
      </ScopedObserver>
    </ScopedObserver>
  </ScopedObserver>
</div>
`;
